src/cornerstone.js

```diff
@@ -0,0 +1,241 @@
+      corpname:  this.corpname
+    });
+
+    return this.auth;
+  }
+
+  async getConnectionSession({httpUrl, method}) {
+
+    const session = await this.auth.setSession({dateUTC: this.dateTime});
+
+    const baseUrl = this.auth.getBaseUrl({corpname: this.corpname});
+    console.log('[setConnectionSession] - baseUrl: ', baseUrl);
+    console.log('[setConnectionSession] - httpUrl: ', httpUrl);
+
+    const sessionSignature = this.auth.getSignatureSession({
+      method:        method,
+      sessionToken:  session.token,
+      sessionSecret: session.secret,
+      httpUrl:       httpUrl,
+      dateUTC:       this.dateTime
+    });
+
+    console.log('[setConnectionSession] - signature: ', sessionSignature);
+    return await this.auth.setConnectionSession({
+      baseUrl:   baseUrl,
+      dateUTC:   this.dateTime,
+      token:     session.token,
+      signature: sessionSignature
+    });
+  }
+
+  async getReporting({request, url}) {
+    this.setAuth();
+    const path = this.auth.getBaseUrl({corpname: this.corpname}) + url + request;
+    console.log('[getReporting] - path: ', path);
+    const connectionSession = await this.getConnectionSession({
+      httpUrl: url,
+      method:  'GET'
+    });
+
+    try {
+      const userObject = await connectionSession.get(path);
+      if (userObject.status === 200 || userObject.status === 201) {
+        console.log('[getReporting] - response: ', userObject.data);
+        return userObject.data.value[0];
+      }
+    } catch (e) {
+      console.log('[getReporting] - Error:', e.response.data);
+    }
+    console.log('erere');
+    return null;
+  }
+
+  async getReportingUserByEmail({email}) {
+    const request = "?$filter=user_email eq '{email}'".replace('{email}', email);
+    const url     = config.CORNERSTONE_REPORTING_USER;
+
+    return await this.getReporting({request: request, url});
+  }
+
+  async getReportingByUserId({user_id}) {
+    const request = "?$filter=user_id eq {user_id}".replace('{user_id}', user_id);
+    const url     = config.CORNERSTONE_REPORTING_USER;
+
+    return await this.getReporting({request: request, url: url});
+  }
+
+  async getReportingByUserRef({user_ref}) {
+    const request = "?$filter=user_ref eq '{user_ref}'".replace('{user_ref}', user_ref);
+    const url     = config.CORNERSTONE_REPORTING_USER;
+
+    return await this.getReporting({request: request, url: url});
+  }
+
+  async getReportingKeycodeByUserId({user_id}) {
+    const request = "?$filter=tu_contact_user_id eq {user_id}".replace('{user_id}', user_id);
+    const url     = config.CORNERSTONE_REPORTING_KEYCODE;
+
+    return await this.getReporting({request: request, url: url});
+  }
+
+  async getReportingKeycodeByUserRef({user_ref}) {
+    const request = "?$filter=tu_training_unit_key_code eq '{user_ref}'".replace('{user_ref}', user_ref);
+    const url     = config.CORNERSTONE_REPORTING_KEYCODE;
+
+    return await this.getReporting({request: request, url: url});
+  }
```

Comment:
There's a lot of duplicated code that could be moved to `getReporting`.
Ex:
```js
async getReportingKeycodeByUserRef({user_ref}) {
  return await this.getReporting({
    url: config.CORNERSTONE_REPORTING_KEYCODE
    filters: [{ tu_training_unit_key_code: { eq: user_ref } }], 
  });
}
```

Reply:
the requests  are not only filter, there are more possibility.

Reply:
Show me where :)